### PR TITLE
Add c{float,double}_fma function in pyopencl-complex.h

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by isuruf** [comment]
This is useful for generating optimized fma in loopy.

File: pyopencl/cl/pyopencl-complex.h [hunk 1]
```diff
@@ -77,6 +77,13 @@
     ; \
   } \
   \
+  inline TP TPROOT##_fma(TP c, TP a, TP b)            \
+  { \
+    return TPROOT##_new( \
+        (c.real + a.real*b.real) - a.imag*b.imag, \
+        (c.imag + a.real*b.imag) + a.imag*b.real); \
+  } \
+  \
```
Comment (inducer) [inline pyopencl/cl/pyopencl-complex.h]: That's sort of false advertisement. :) This doesn't use `fma` internally.

Comment (inducer) [inline pyopencl/cl/pyopencl-complex.h]: Want to add a test, or at least something that exercises this functionality?

**Review by inducer — COMMENTED** [review]
Thanks! Two minor wrinkles below.

File: pyopencl/cl/pyopencl-complex.h (hunk 1 above)
Comment (isuruf) [reply pyopencl/cl/pyopencl-complex.h]: I renamed to `c{double, float}_muladd` to not say that this is a fused multiply-add.

Comment (isuruf) [reply pyopencl/cl/pyopencl-complex.h]: Done in 6d5bbf4

Comment (inducer) [reply pyopencl/cl/pyopencl-complex.h]: That's not how I figured this might go. Why not just use the FMA function in OpenCL?

**Comment by inducer** [comment]
LGTM. Thanks!

File: pyopencl/cl/pyopencl-complex.h [hunk 2]
```diff
@@ -77,6 +77,13 @@
     ; \
   } \
   \
+  inline TP TPROOT##_fma(TP a, TP b, TP c)           \
+  { \
+    return TPROOT##_new( \
+        fma(a.real, b.real, c.real) - a.imag*b.imag, \
```
Comment (alexfikl) [inline pyopencl/cl/pyopencl-complex.h:83]: @isuruf Just curious, why isn't this?
```c
fma(fma(a.real, b.real, c.real), -a.imag, b.imag)
```

Comment (isuruf) [reply pyopencl/cl/pyopencl-complex.h:83]: A NEG and an FMA costs the same as a MUL and a SUB.
Also, this way the compiler is free to determine if
```
fma(fma(a.real, b.real, c.real), -a.imag, b.imag)
```
or
```
fma(fma(a.real, b.real, c.real), a.imag, -b.imag)
```
is better.

Comment (alexfikl) [reply pyopencl/cl/pyopencl-complex.h:83]: Ah, cool! Thanks for the explanation!

**Pull request merged**